Mathlib/Algebra/Group/AddChar.lean

```diff
@@ -85,49 +87,66 @@ open Multiplicative
 /-- Define coercion to a function so that it includes the move from `A` to `Multiplicative A`.
 After we have proved the API lemmas below, we don't need to worry about writing `ofAdd a`
 when we want to apply an additive character. -/
-instance instFunLike : FunLike (AddChar A M) A M :=
+instance instFunLike [Monoid M] : FunLike (AddChar A M) A M :=
   inferInstanceAs (FunLike (Multiplicative A →* M) A M)
 #noalign add_char.has_coe_to_fun
 
-theorem coe_to_fun_apply (ψ : AddChar A M) (a : A) : ψ a = ψ.toMonoidHom (ofAdd a) :=
+theorem coe_to_fun_apply [Monoid M] (ψ : AddChar A M) (a : A) : ψ a = ψ.toMonoidHom (ofAdd a) :=
   rfl
 #align add_char.coe_to_fun_apply AddChar.coe_to_fun_apply
 
 -- Porting note: added
-theorem mul_apply (ψ φ : AddChar A M) (a : A) : (ψ * φ) a = ψ a * φ a :=
+theorem mul_apply [CommMonoid M] (ψ φ : AddChar A M) (a : A) : (ψ * φ) a = ψ a * φ a :=
   rfl
 
 -- Porting note: added
 @[simp]
-theorem one_apply (a : A) : (1 : AddChar A M) a = 1 := rfl
+theorem one_apply [CommMonoid M] (a : A) : (1 : AddChar A M) a = 1 := rfl
 
 -- this instance was a bad idea and conflicted with `instFunLike` above
 #noalign add_char.monoid_hom_class
 
 -- Porting note(https://github.com/leanprover-community/mathlib4/issues/5229): added.
 @[ext]
-theorem ext (f g : AddChar A M) (h : ∀ x : A, f x = g x) : f = g :=
+theorem ext [Monoid M] (f g : AddChar A M) (h : ∀ x : A, f x = g x) : f = g :=
   MonoidHom.ext h
 
 /-- An additive character maps `0` to `1`. -/
 @[simp]
-theorem map_zero_one (ψ : AddChar A M) : ψ 0 = 1 := by rw [coe_to_fun_apply, ofAdd_zero, map_one]
+theorem map_zero_one [Monoid M] (ψ : AddChar A M) : ψ 0 = 1 := by
+  rw [coe_to_fun_apply, ofAdd_zero, map_one]
 #align add_char.map_zero_one AddChar.map_zero_one
 
 /-- An additive character maps sums to products. -/
 @[simp]
-theorem map_add_mul (ψ : AddChar A M) (x y : A) : ψ (x + y) = ψ x * ψ y := by
+theorem map_add_mul [Monoid M] (ψ : AddChar A M) (x y : A) : ψ (x + y) = ψ x * ψ y := by
   rw [coe_to_fun_apply, coe_to_fun_apply _ x, coe_to_fun_apply _ y, ofAdd_add, map_mul]
 #align add_char.map_add_mul AddChar.map_add_mul
 
 /-- An additive character maps multiples by natural numbers to powers. -/
 @[simp]
-theorem map_nsmul_pow (ψ : AddChar A M) (n : ℕ) (x : A) : ψ (n • x) = ψ x ^ n := by
+theorem map_nsmul_pow [Monoid M] (ψ : AddChar A M) (n : ℕ) (x : A) : ψ (n • x) = ψ x ^ n := by
   rw [coe_to_fun_apply, coe_to_fun_apply _ x, ofAdd_nsmul, map_pow]
 #align add_char.map_nsmul_pow AddChar.map_nsmul_pow
 
 end CoeToFun
 
+section Constructor
+
+/-- Construct an `AddChar` from a function satisfying `f (a + b) = f a * f b` and `f 0 = 1`. -/
```

Comment:
Why do you want `f` implicit here? Don't you think that a call like `mk f` would be useful?

Reply:
You're right – I realised this when I went on to refactor some existing code, we need `f` explicit to be able to write `AddChar.mk (some_function) (by simp) (by simp)`